openml/datasets/dataset.py

```diff
@@ -208,15 +206,31 @@ def remove_tag(self, tag):
         openml._api_calls._perform_api_call("/data/untag", data=data)
 
     def __eq__(self, other):
+
+        server_fields = {
+            'dataset_id': True,
+            'version': True,
+            'upload_date': True,
+            'url': True,
+            'dataset': True,
+            'data_file': True,
+        }
+
         if type(other) != OpenMLDataset:
             return False
-        elif (
-            self.dataset_id == other.dataset_id
-            or (self.name == other._name and self.version == other._version)
-        ):
-            return True
         else:
-            return False
+            for field in self.__dict__:
+                if field not in server_fields:
+                    if field in other.__dict__:
+                        if self.__dict__[field] != other.__dict__[field]:
+                            return False
+                    else:
```

Comment:
I personnally find that the following will be more readable:

```python
        server_fields = {'dataset_id', 'version', 'upload_data', 'url',
                         'dataset', 'data_file'}

        if type(other) != OpenMLDataset:
            return False

        # check that the keys are identical
        self_keys = set(self.__dict__.keys()).difference_update(server_fields)
        other_keys = set(other.__dict__.keys()).difference_update(server_fields)
        if self_keys != other_keys:
            return False

        # check that values of the common keys are identical
        return all(self.__dict__[key] == other.__dict__[key]
                   for key in self_keys)
```

Reply:
I am not sure about this one, it looks the other way around to me.

Reply:
so, the reason that I think this is not good is that supposing that the object has 'size' fields. 
And that the server fields are of a size t.
`self_keys = set(self.__dict__.keys()).difference_update(server_fields) `
for this you have a complexity of 't'
`other_keys = set(other.__dict__.keys()).difference_update(server_fields)`
for this you have the complexity too
```
 if self_keys != other_keys:
            return False
```
Taking a guess and saying this has 'size-t' complexity checking if size is equal and going through each element.
```
return all(self.__dict__[key] == other.__dict__[key]
                   for key in self_keys)
```
You iterate over each key left after removing the server fields, complexity 'size-t'.
You then call all which still iterates over each item on the iterable, complexity 'size-t'.
So in the end your code has 'size+ size+ size -t complexity'

I might be wrong on my analysis, but I would say my code has a 'size' complexity and if there is a field which is not equal it stops right away, yours goes until the end.

Reply:
```
if self_keys != other_keys:
            return False
```
I think this would be better expressed as
```
if len(self_keys) != len(other_keys):
            return False
```